File: Hawkeye-Python code/hawkeye_test.py
import math
import random
import numpy as np
from helper_function import CRC, History
from optgen import OPTgen, OPTGEN_SIZE
from predictor import Hawkeye_Predictor
from hawkeye_algorithm import UpdateReplacementState,InitReplacementState,GetVictimInSet,LLC_SETS, optgen_occup_vector,LLC_SETS, optgen_occup_vector, print_stats_heartbeat,print_stats

with open('VictimWays.txt', "w") as file:
    pass

####################################################################################################
## _____      ___       __   _____
##   |       |___      (__     |
##   |       |___       __)    |
###################################################################################################


#Test the algorithm with num_accesses that will be generate in the file

# np.random.seed(42)
# # random.seed(seed_value)
# num_accesses = 500
# cache_accesses = []

# for _ in range(num_accesses):
#     set_num = random.randint(0, LLC_SETS -1 )  # LLC_SETS = 2048
#     way_num = random.randint(0, LLC_WAYS -1)  # LLC_WAYS = 16
#     access_type = random.choice(["PREFETCH", "DEMAND"])
#     is_hit = random.choice([True, False])
#     paddr = random.randint(0x1000, 0xFFFF)
#     PC = random.randint(0x1000, 0xFFFF)
#     cache_accesses.append(f"{set_num}, {way_num}, {access_type}, {is_hit}, 0x{paddr:X}, 0x{PC:X}")

# # # Join the cache accesses into a single string
# cache_access_sequence = "\n".join(cache_accesses)


# Read cache accesses from the file
with open('/mnt/e/F3CAS/Hawkeye/BeladyCacheReplacement/cache_accesses.txt', "r") as file:
    cache_access_lines = file.readlines()

# write cache accesses in the file
# with open('cache_accesses.txt', "w") as file:
#     file.write(cache_access_sequence)



# def simulate_cache_access(set_number, way_number, access_type, is_hit, paddr, PC):
#     """
#     Simulates a cache access and updates the replacement state accordingly.

#     :param set_number: The cache set number.
#     :param way_number: The cache way number.
#     :param access_type: Type of access, either "PREFETCH" or "DEMAND".
#     :param is_hit: Boolean indicating if the access is a hit (True) or miss (False).
#     :param paddr: The physical address being accessed.
#     :param PC: The program counter value for the access.
#     """
#     victim_addr = 0  # Assuming 0, "as default victim address"
#     UpdateReplacementState(0, set_number, way_number, paddr, PC, victim_addr, access_type, int(is_hit))

def print_cache_state(set_number):
    """
    Prints the state of the cache for a specific set.

    :param set_number: The cache set number to print the state for.
    """
    # print(f"Set {set_number} State:")
    # print("RRIP:", rrip[set_number])
    # print("Prefetching:", prefetching[set_number])
    # print("Sample Signature:", sample_signature[set_number])
    # print("Set Timer:", set_timer[set_number])
    # print()


# cache_access_sequence = [
#     (0, 0, "PREFETCH", False, 0x1000, 0x2000),
#     (0, 0, "PREFETCH", True, 0x1000, 0x2000),
#     (0, 0, "DEMAND", False, 0x1000, 0x2000),
#     (0, 0, "PREFETCH", False, 0x1000, 0x2000),
#     (0, 0, "PREFETCH", False, 0x1000, 0x4000),
#     (0, 0, "DEMAND", False, 0x1000, 0x4000),
#     (0, 0, "PREFETCH", False, 0x1000, 0x2000),
#     (0, 0, "PREFETCH", True, 0x1000, 0x2000),
#     (0, 0, "DEMAND", False, 0x1000, 0x2000),
#     (0, 0, "PREFETCH", False, 0x1000, 0x2000),
#     (0, 0, "PREFETCH", False, 0x1000, 0x4000),
#     (0, 0, "PREFETCH", False, 0x1000, 0x4000),
#     (0, 0, "DEMAND", True, 0x1000, 0x4000),
#     (0, 0, "DEMAND", False, 0x1000, 0x4000),
#     (0, 0, "DEMAND", True, 0x1000, 0x4000),
#     (0, 0, "DEMAND", False, 0x1000, 0x4000),
#     # ... add more accesses as needed
# ]

# Initialize the replacement state
InitReplacementState()

# # Simulate the defined sequence of cache accesses - without file
# for access in cache_access_sequence:
#     set_number, way_number, access_type, is_hit, paddr, PC = access
#     print(f"Simulating Cache Access: Set={set_number}, Way={way_number}, Type={access_type}, Hit={is_hit}, Addr={hex(paddr)}, PC={hex(PC)}")
#     simulate_cache_access(set_number, way_number, access_type, is_hit, paddr, PC)
#     victim_way = GetVictimInSet(0, set_number, 0, PC, paddr, access_type)
#     print("victim_way:", victim_way)
#     print_cache_state(set_number)


# Simulate the defined sequence of cache accesses
for access_line in cache_access_lines:
    access_line = access_line.strip()  # Remove newline character
    if not access_line:
        continue  # Skip empty lines

    # Split the line into components
    access_components = access_line.split(", ")
    # if len(access_components) != 6:
    #     print(f"Skipping invalid line: {access_line}")
    #     continue

    # Unpack the components
    set_number, way_number, access_type, is_hit, paddr_str, PC_str = access_components
    set_number = np.uint32(set_number)
    way_number = np.uint32(way_number)
    access_type = np.uint32(access_type)
    is_hit = is_hit == "True"
    paddr = np.uint32(paddr_str)  # Convert hexadecimal string to integer
    PC = np.uint32(PC_str)          # Convert hexadecimal string to integer

    # print(f"Simulating Cache Access: Set={set_number}, Way={way_number}, Type={access_type}, Hit={is_hit}, Addr={hex(paddr)}, PC={hex(PC)}")
   
    victim_way = GetVictimInSet(0, set_number, 0, PC, paddr, access_type)
    UpdateReplacementState(0, set_number, victim_way, paddr, PC, 0, access_type, is_hit)
    # simulate_cache_access(set_number, victim_way, access_type, is_hit, paddr, PC)
    
    # print("Victim way:", victim_way)
    sequential_mapping = []
    mapping_str = f"0x{paddr} -> VictimWay: {victim_way}, set: {set_number}"
    sequential_mapping.append(mapping_str)

    with open('VictimWays.txt', "a") as file:
        for mapping in sequential_mapping:
            file.write(mapping + "\n")
    print_cache_state(set_number)


# # Simulate the defined sequence of cache accesses
# cache_access_lines = cache_access_sequence.split("\n")
# for access_line in cache_access_lines:
#     # Split the line into components
#     access_components = access_line.split(", ")

#     # Unpack the components
#     set_number, way_number, access_type, is_hit, paddr_str, PC_str = access_components
#     set_number = int(set_number)
#     way_number = int(way_number)
#     is_hit = is_hit == "True"
#     paddr = int(paddr_str, 16)  # Convert hexadecimal string to integer
#     PC = int(PC_str, 16)        # Convert hexadecimal string to integer

#     print(f"Simulating Cache Access: Set={set_number}, Way={way_number}, Type={access_type}, Hit={is_hit}, Addr={hex(paddr)}, PC={hex(PC)}")
#     simulate_cache_access(set_number, way_number, access_type, is_hit, paddr, PC)
#     victim_way = GetVictimInSet(0, set_number, 0, PC, paddr, access_type)
#     print("Victim way:", victim_way)
#     print_cache_state(set_number)

# Print final statistics
print_stats_heartbeat(LLC_SETS, optgen_occup_vector)
print_stats(LLC_SETS, optgen_occup_vector)

File: Hawkeye-CPP-test codes/hawkeye_test.cc
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <ctime>
#include <random>
#include "hawkeye_algorithm.cc"  // Include your Hawkeye algorithm header file

#define DEMAND 0
// const uint32_t PREFETCH = 1;  // Define the PREFETCH constant

struct CacheAccess {
    uint32_t set;
    uint32_t way;
    uint32_t type;
    bool is_hit;
    uint64_t paddr;
    uint64_t PC;
};

// void simulate_cache_access(const CacheAccess& access) {
//     UpdateReplacementState(0, access.set, access.way, access.paddr, access.PC, 0, access.type, access.is_hit);
// }


void print_cache_state(uint32_t set) {
    // Implement this based on your Hawkeye algorithm structure
}

// CacheAccess parse_access(const std::string& line) {
//     std::istringstream iss(line);
//     CacheAccess access;
//     char comma;
//     std::string accessType;

//     iss >> access.set >> comma >> access.way >> comma >> accessType >> comma 
//         >> access.is_hit >> comma >> std::hex >> access.paddr >> comma >> access.PC;
//     std::dec(iss); // Switch back to decimal mode for subsequent reads

//     access.type = (accessType == "PREFETCH") ? PREFETCH : DEMAND;

//     return access;
// }

// int main() {
//     std::string filePath = "/mnt/e/F3CAS/Python-code/Hawkeye/cache_accesses.txt";
//     std::ifstream file(filePath);
//     std::string line;
//     std::vector<CacheAccess> cache_access_sequence;

//     if (!file) {
//         std::cerr << "Unable to open file at " << filePath << std::endl;
//         return 1;
//     }

//     while (std::getline(file, line)) {
//         cache_access_sequence.push_back(parse_access(line));
//     }

//     InitReplacementState();

//     for (const auto& access : cache_access_sequence) {
//         std::cout << "Simulating Cache Access: Set=" << access.set 
//                   << ", Way=" << access.way 
//                   << ", Type=" << (access.type == PREFETCH ? PREFETCH : DEMAND)
//                   << ", Hit=" << (access.is_hit ? "True" : "False") 
//                   << ", Addr=" << std::hex << access.paddr 
//                   << ", PC=" << std::dec << access.PC << std::endl;

//         simulate_cache_access(access);
//         print_cache_state(access.set);
//     }

//     // Replace these with the correct function calls if available.
//     PrintStats_Heartbeat();
//     PrintStats();

//     return 0;
// }


// int main() {
//     // std::string filePath = "/mnt/e/F3CAS/Python-code/Hawkeye/cache_accesses.txt";
//     std::ifstream inFile("/mnt/e/F3CAS/Python-code/Hawkeye/cache_accesses.txt");
//     if (!inFile.is_open()) {
//             std::cerr << "Failed to open the file for reading." << std::endl;
//             return 1;
//         }


//     InitReplacementState();
//     std::string line;
//     while (std::getline(inFile, line)) {
//         std::istringstream iss(line);
//         uint32_t set, way;
//         std::string accessType, hit;
//         uint64_t paddr, PC;

//         iss >> set >> way;
//         iss.ignore(2); // Ignore the comma and space
//         iss >> accessType >> hit >> std::hex >> paddr >> std::hex >> PC;

//         bool is_hit = (hit == "True") ? true : false;
//         uint32_t type = (accessType == "PREFETCH") ? PREFETCH : DEMAND;

//         std::cout << "Simulating Cache Access: Set=" << set 
//                   << ", Way=" << way 
//                   << ", Type=" << (accessType == "PREFETCH" ? PREFETCH : DEMAND)
//                   << ", Hit=" << (is_hit ? "True" : "False") 
//                   << ", Addr=" << std::hex << paddr 
//                   << ", PC=" << std::dec << PC << std::endl;

//         simulate_cache_access(set, way, type, is_hit, paddr, PC);
//         print_cache_state(set);
//     }

//     inFile.close();

//     // Replace these with the correct function calls if available.
//     PrintStats_Heartbeat();
//     PrintStats();

//     return 0;
// }

int main() {
    // Seed the random number generator
    std::mt19937 generator(42);
    std::uniform_int_distribution<uint32_t> set_distribution(0, LLC_SETS -1 );  // Assuming LLC_SETS = 2048
    std::uniform_int_distribution<uint32_t> way_distribution(0, LLC_WAYS -1 );    // Assuming LLC_WAYS = 16
    std::uniform_int_distribution<uint32_t> type_distribution(0, 3);    // 0 for DEMAND, 2 for PREFETCH, 3 for WRITEBACK
    std::uniform_int_distribution<uint32_t> hit_distribution(0, 1);     // 0 for False, 1 for True
    std::uniform_int_distribution<uint64_t> addr_distribution(0x1000, 0xFFFF);
     std::uniform_int_distribution<uint64_t> PC_distribution(0x1000, 0xFFFF);

    // Open a file for writing
    std::ofstream outFile("cache_accesses.txt");
    if (!outFile.is_open()) {
        std::cerr << "Failed to open the cache_accesses for writing." << std::endl;
        return 1;
    }
    // VictimWay file
        // Open a file for writing
    std::ofstream VictimFile("VictimWays.txt");
    if (!VictimFile.is_open()) {
        std::cerr << "Failed to open the file for writing." << std::endl;
        return 1;
    }

    InitReplacementState();

    // Generate and write random cache accesses
    for (int i = 0; i < 9000; ++i) {
        CacheAccess access;
        access.set = set_distribution(generator);
        access.way = way_distribution(generator);
        access.type = type_distribution(generator);     // == 0 ? DEMAND : PREFETCH;
        access.is_hit = hit_distribution(generator) == 0 ? false : true;
        access.paddr = addr_distribution(generator);
        access.PC = PC_distribution(generator);

        // Write to file
        outFile << access.set << ", " << access.way << ", " 
                << access.type << ", " 
                << (access.is_hit ? "True" : "False") << ", " 
                << access.paddr << ", " << access.PC  << "\n";

        // Print to console
        // std::cout << "Simulating Cache Access: Set=" << access.set 
        //           << ", Way=" << access.way 
        //           << ", Type=" << (access.type == PREFETCH ? "PREFETCH" : "DEMAND")
        //           << ", Hit=" << (access.is_hit ? "True" : "False") 
        //           << ", Addr=0x" << std::hex << access.paddr 
        //           << ", PC=0x" << access.PC << std::dec << std::endl;

        // simulate_cache_access(access);
        
        print_cache_state(access.set);
        int victim_way = GetVictimInSet(0, access.set, 0, access.PC, access.paddr, access.type);
        UpdateReplacementState(0, access.set, victim_way, access.paddr, access.PC, 0, access.type, access.is_hit);
        // //SM
        // cout << "rrip array " << endl;
        // for (uint32_t i = 0; i < LLC_WAYS; i++) {
        //     cout << rrip[access.set][i] << " ";
        // }
        // cout << endl;  // Newline for each set
   
        //  //endSM
        // cout<<"Victim way:"<< victim_way<<endl;
        VictimFile << "0x"<< access.paddr << " -> " << "VictimWay: " << victim_way << ", set: " << access.set <<endl;
        
    }

    // Close the file
    outFile.close();
    VictimFile.close();

    std::cout << "Cache access sequence has been written to cache_accesses.txt" << std::endl;

    PrintStats_Heartbeat();
    PrintStats();

    return 0;
}
// int main() {
//     std::vector<CacheAccess> cache_access_sequence = {
//         {0, 0, PREFETCH, false, 0x1000, 0x2000},
//         {0, 0, PREFETCH, true, 0x1000, 0x2000},
//         {0, 0, DEMAND, false, 0x1000, 0x2000},
//         {0, 0, PREFETCH, false, 0x1000, 0x2000},
//         {0, 0, PREFETCH, false, 0x1000, 0x4000},
//         {0, 0, DEMAND, false, 0x1000, 0x4000},
//         {0, 0, PREFETCH, false, 0x1000, 0x2000},
//         {0, 0, PREFETCH, true, 0x1000, 0x2000},
//         {0, 0, DEMAND, false, 0x1000, 0x2000},
//         {0, 0, PREFETCH, false, 0x1000, 0x2000},
//         {0, 0, PREFETCH, false, 0x1000, 0x4000},
//         {0, 0, PREFETCH, false, 0x1000, 0x4000},
//         {0, 0, DEMAND, true, 0x1000, 0x4000},
//         {0, 0, DEMAND, false, 0x1000, 0x4000},
//         {0, 0, DEMAND, true, 0x1000, 0x4000},
//         {0, 0, DEMAND, false, 0x1000, 0x4000},
//         // ... add more accesses as needed
//     };

//     InitReplacementState();

//     for (const auto& access : cache_access_sequence) {
//         std::cout << "Simulating Cache Access: Set=" << access.set 
//                   << ", Way=" << access.way 
//                   << ", Type=" << (access.type == PREFETCH ? "PREFETCH" : "DEMAND")
//                   << ", Hit=" << (access.is_hit ? "true" : "false") 
//                   << ", Addr=" << std::hex << access.paddr 
//                   << ", PC=" << std::dec << access.PC << std::endl;

//         simulate_cache_access(access);
//         print_cache_state(access.set);
//     }

//     // Replace these with the correct function calls if available.
//     PrintStats_Heartbeat();
//     PrintStats();

//     return 0;
// }
